### ISSUE TITLE
Support for tsconfig.json in project root

### ISSUE THREAD

**Issue by vojtechhabarta** [Everlaw/nbts#21]
Currently TypeScript plugin assumes that TypeScript projects correspond to NetBeans project _source roots_ which I think is usually `src/main/java` for `jar` Maven project and also `src/main/webapp` for `war` project. These source roots can have `tsconfig.json`.

But usually my TypeScript project corresponds to Maven project, I have `tsconfig.json` in  the same directory as `pom.xml`. Currently this project structure is not supported and `tsconfig.json` is ignored.

**Comment by jeffrey-easyesi** [comment]
NetBeans's indexing API only notifies of changes to files that are under a source root. I don't know of any good way around this.

**Comment by stunaz** [comment]
Would be great to have the options as the following project : 
https://github.com/GeertjanWielenga/TypeScript ((b) Extension to Options window) .

**Comment by stunaz** [comment]
Would have been great to configure the location of the tsconfig.json and the tsd.json for the project.

**Comment by RomkeVdMeulen** [comment]
I may have found a workaround for Maven projects (also relevant to #38 and #28): if you add the folder holding your TypeScript sources as a "resource" to your `pom.xml` (but excluding everything so that it doesn't get packaged), then nbts seems to kick in. Not sure of this fits your scenarios exactly, but it may help you find a solution.

For example, to add dir `ts` as a resource without packaging it, add this to `<resources>` in your `pom.xml`:

```
<resource>
	<directory>ts/</directory>
	<excludes>
		<exclude>**/*</exclude>
	</excludes>
</resource>
```

If your `pom.xml` doesn't have a `<resources>` element yet, add it to `<build>` with this content to preserve default behaviour:

```
<resources>
	<resource>
		<directory>src/main/resources</directory>
		<includes>
			<include>**/*</include>
		</includes>
	</resource>
</resources>
```

**Comment by Rycochet** [comment]
I've got something similar in scope so posting under this rather than creating a new issue - shouldn't it try to find tsconfig "somewhere higher in the tree" than the file you're editing rather than a single place?

In my case I've got one project nested inside another, but it only finds the outer project and not the inner one, so there's no code-completion as that tsconfig is ignored (specifically the inner project needs to be standalone but also available as a built module within the outer one).

**Comment by ForNeVeR** [comment]
@jeffrey-easyesi 
> NetBeans's indexing API only notifies of changes to files that are under a source root. I don't know of any good way around this.

Are we really really need that? Couldn't the plugin work without these notifications? You know, that's not the most often activity in the world to change `tsconfig.json`. Couldn't the plugin work with outdated config model? That would be preferable I think.

Personally I tend to place my TypeScript sources into `src/main/ts` directory, and it doesn't seem to be indexed by Netbeans at all. Is there a way around this?

**Comment by everflux** [comment]
I added this issue to NetBeans as enhancement, I have similar problems with the rigid source root configuration. Feel free to comment/vote: https://netbeans.org/bugzilla/show_bug.cgi?id=262747

**Comment by jeffrey-easyesi** [comment]
The latest release, 2.5.3.0, walks up the directory tree looking for a tsconfig.json, so tsconfig.json no longer has to be in a source directory.

**Comment by Rycochet** [comment]
Just a quick question - if you have a project with multiple folders and multiple tsconfig.json files will it treat them separately? Related to that, if you have a project does it stop checking within the folder tree, or does it carry on looking further up?

**Comment by jeffrey-easyesi** [comment]
>if you have a project with multiple folders and multiple tsconfig.json files will it treat them separately?

Yes, mostly. The main thing that's still tied to source roots is error checking, i.e. if you save a file in one of the folders, all files under the source root will get re-checked.

>Related to that, if you have a project does it stop checking within the folder tree, or does it carry on looking further up?

It doesn't know what the project root is, so it'll keep going up to the filesystem root. (This is the same thing `tsc` does when run from the command line without any parameters.)
